penguin/web/sse_events.py

```
@@ -102,11 +108,23 @@ async def events_sse(
         queue: asyncio.Queue[dict] = asyncio.Queue(maxsize=_SSE_QUEUE_MAX_EVENTS)
         # Reconnect-only dedupe bridge: events can arrive after subscription but
-        # before replay drains. The set is intentionally per connection and
-        # should disappear when Phase 11.5 moves replay/dedupe to the ledger.
+        # before replay drains. The set is intentionally per connection; durable
+        # replay identity comes from the runtime event ledger.
         queued_live_event_ids: set[str] = set()
+        delivered_event_ids: set[str] = set()
+        delivered_event_order: deque[str] = deque()
         event_order = 0
 
+        def mark_delivered(event_id: Any) -> None:
```

Ruff (0.15.20)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `event_id`

(ANN401)